awx/api/serializers.py

```diff
@@ -2352,49 +2386,50 @@ def to_representation(self, obj):
             ret['project'] = None
             if 'playbook' in ret:
                 ret['playbook'] = ''
-        if 'credential' in ret and not obj.credential:
-            ret['credential'] = None
-        if 'vault_credential' in ret and not obj.vault_credential:
-            ret['vault_credential'] = None
-        if self.version == 1 and 'credential' in self.Meta.fields:
+        ret['credential'] = obj.credential
+        ret['vault_credential'] = obj.vault_credential
+        if self.version == 1:
             ret['cloud_credential'] = obj.cloud_credential
             ret['network_credential'] = obj.network_credential
         return ret
 
     def create(self, validated_data):
         deprecated_fields = {}
-        for key in ('cloud_credential', 'network_credential'):
+        for key in ('credential', 'vault_credential', 'cloud_credential', 'network_credential'):
             if key in validated_data:
                 deprecated_fields[key] = validated_data.pop(key)
         obj = super(JobOptionsSerializer, self).create(validated_data)
-        if self.version == 1 and deprecated_fields:  # TODO: remove in 3.3
+        if deprecated_fields:  # TODO: remove in 3.3
             self._update_deprecated_fields(deprecated_fields, obj)
         return obj
 
     def update(self, obj, validated_data):
         deprecated_fields = {}
-        for key in ('cloud_credential', 'network_credential'):
+        for key in ('credential', 'vault_credential', 'cloud_credential', 'network_credential'):
             if key in validated_data:
                 deprecated_fields[key] = validated_data.pop(key)
         obj = super(JobOptionsSerializer, self).update(obj, validated_data)
-        if self.version == 1 and deprecated_fields:  # TODO: remove in 3.3
+        if deprecated_fields:  # TODO: remove in 3.3
             self._update_deprecated_fields(deprecated_fields, obj)
         return obj
 
     def _update_deprecated_fields(self, fields, obj):
         for key, existing in (
+            ('credential', obj.credentials.filter(credential_type__kind='ssh')),
+            ('vault_credential', obj.credentials.filter(credential_type__kind='vault')),
             ('cloud_credential', obj.cloud_credentials),
             ('network_credential', obj.network_credentials),
         ):
             if key in fields:
                 for cred in existing:
-                    obj.extra_credentials.remove(cred)
+                    obj.credentials.remove(cred)
                 if fields[key]:
-                    obj.extra_credentials.add(fields[key])
+                    obj.credentials.add(fields[key])
```

Comment:
okay, this answers my prior question about how `credential` gets both removed and added.

Reply:
Yep, same as net and cloud - if you specify the deprecated flag, we remove _all_ of the existing creds that match that type.  In other words, if you specify `credential=` it enforces that as the _only_ machine/SSH in the list.  If you specify `vault_credential=` it enforces that as the _only_ vault in the list.

Reply:
How is RBAC handled for this? The access methods are not equipped to process either non-field keys like "credential" or many-to-many fields like "credentials". Even if they were, the code that calls the access method wouldn't know how to pass it.

Scenario in question: someone PATCHes a job template, providing `credential`. Does their "use" access to the credential get checked?

The most canonical answer I know of is that we check RBAC in the view, before a particular action takes place. That's not going to be easy to get to work with the implementation you've done here.

Reply:
Yea, that's a good point.  I may need an explicit `use_role` check here since we're sort of cheating and short-circuiting attach/unattach.

Reply:
I found where we do that. It's `.validate` on `JobOptions`

https://github.com/ryanpetrello/awx/blob/6d9d5a52c32e6439ecca676ca7c97047c97c05a6/awx/api/serializers.py#L2435

there we go.

Reply:
Perfect.  I'll replicate the same behavior here:

https://github.com/ryanpetrello/awx/blob/6d9d5a52c32e6439ecca676ca7c97047c97c05a6/awx/api/serializers.py#L2456

Reply:
Addressed via:

https://github.com/ansible/awx/pull/595/files#diff-a81324c523b41de7296fdd5ff9063d10R2432